### PR TITLE
#51 Sonar fixes for "When the "Rule-of-Zero" is not applicable, the "…

### PR DESCRIPTION
**Pull request by rosiereflo**
…Rule-of-Five" should be followed" and "Floating point numbers should not be tested for equality"

### CODE REVIEW

File: dynawo/sources/API/CRV/CRVCurve.cpp [hunk 1]
```diff
@@ -39,6 +39,16 @@ Curve::const_iterator::const_iterator(const Curve::const_iterator& original) {
...
 Curve::const_iterator::~const_iterator() {
   delete impl_;
+  impl_ = NULL;
+}
+
+Curve::const_iterator&
+Curve::const_iterator::operator=(const Curve::const_iterator& other) {
+  if (this == &other)
+    return *this;
+  delete impl_;
```
Comment (gautierbureau) [inline dynawo/sources/API/CRV/CRVCurve.cpp]: if ( impl_ != NULL ) delete impl_; ?

File: dynawo/sources/API/CRV/CRVCurve.cpp [hunk 2]
```diff
@@ -39,6 +39,16 @@ Curve::const_iterator::const_iterator(const Curve::const_iterator& original) {
...
+  impl_ = new BaseConstIteratorImpl(*(other.impl_));
```
Comment (gautierbureau) [inline dynawo/sources/API/CRV/CRVCurve.cpp]: Test if other.impl_ is not NULL also ?

File: dynawo/sources/Common/DYNCommun.h [hunk 3]
```diff
@@ -87,6 +90,31 @@ namespace DYN {
...
+  /**
+   * @brief return true if a == b
+   *
+   *
+   * @param a first double
+   * @param b second double
+   * @return return true if a == b
+   */
+  static inline bool doubleEquals(const double& a, const double& b) {
+    return std::abs(a-b) < std::numeric_limits<double>::epsilon() * std::abs(a+b) * 1000
```
Comment (gautierbureau) [inline dynawo/sources/Common/DYNCommun.h]: std::fabs

File: dynawo/sources/Modeler/Common/DYNConnector.cpp [hunk 4]
```diff
@@ -610,7 +602,7 @@ ConnectorContainer::getY0Connector() {
             it != zc->connectedSubModels().end();
             ++it) {
       const int numVar = it->subModel()->getVariableIndexGlobal(it->variable());
-      if (zLocal_[numVar] != 0) {  // non zero variable
+      if (doubleNotEquals(zLocal_[numVar], 0)) {  // non zero variable
```
Comment (gautierbureau) [inline dynawo/sources/Modeler/Common/DYNConnector.cpp]: 0.

File: dynawo/sources/Modeler/DataInterface/DYNLineInterfaceIIDM.cpp [hunk 5]
```diff
@@ -104,7 +105,7 @@ LineInterfaceIIDM::getR() const {
 
 double
 LineInterfaceIIDM::getX() const {
-  if (lineIIDM_.x() == 0 && lineIIDM_.r() == 0) {
+  if (doubleEquals(lineIIDM_.x(), 0) && doubleEquals(lineIIDM_.r(), 0)) {
```
Comment (gautierbureau) [inline dynawo/sources/Modeler/DataInterface/DYNLineInterfaceIIDM.cpp]: 0.

File: dynawo/sources/Modeler/DataInterface/DYNTwoWTransformerInterfaceIIDM.cpp [hunk 6]
```diff
@@ -345,7 +346,7 @@ TwoWTransformerInterfaceIIDM::importStaticParameters() {
   staticParameters_["q2"] = StaticParameter("q2", StaticParameter::DOUBLE).setValue(getQ2());
 
   double i1 = 0;
-  if (getInitialConnected1() && busInterface1_->getV0() != 0) {
+  if (getInitialConnected1() && doubleNotEquals(busInterface1_->getV0(), 0)) {
```
Comment (gautierbureau) [inline dynawo/sources/Modeler/DataInterface/DYNTwoWTransformerInterfaceIIDM.cpp]: 0.

File: dynawo/sources/Modeler/DataInterface/DYNTwoWTransformerInterfaceIIDM.cpp [hunk 7]
```diff
@@ -356,7 +357,7 @@ TwoWTransformerInterfaceIIDM::importStaticParameters() {
   }
 
   double i2 = 0;
-  if (getInitialConnected2() && busInterface2_->getV0() != 0) {
+  if (getInitialConnected2() && doubleNotEquals(busInterface2_->getV0(), 0)) {
```
Comment (gautierbureau) [inline dynawo/sources/Modeler/DataInterface/DYNTwoWTransformerInterfaceIIDM.cpp]: I would even maybe create a IsZero function in DYNCommun...

File: dynawo/sources/Modeler/DataInterface/test/TestIIDMModels.cpp [hunk 8]
```diff
@@ -454,10 +454,10 @@ TEST(DataInterfaceIIDMTest, testDanglingLineIIDMCheckCriteriaAndStaticParameters
   exportStateVariables(data);
 
   ASSERT_EQ(data->checkCriteria(true), true);
-  ASSERT_DOUBLE_EQUALS_DYNAWO(data->getStaticParameterDoubleValue("MyDanglingLine", "p_pu"), 105. / SNREF);
-  ASSERT_DOUBLE_EQUALS_DYNAWO(data->getStaticParameterDoubleValue("MyDanglingLine", "q_pu"), 90. / SNREF);
-  ASSERT_DOUBLE_EQUALS_DYNAWO(data->getStaticParameterDoubleValue("MyDanglingLine", "p"), 105.);
-  ASSERT_DOUBLE_EQUALS_DYNAWO(data->getStaticParameterDoubleValue("MyDanglingLine", "q"), 90.);
+  ASSERT_DOUBLE_EQUALS_DYNAWO(data->getStaticParameterDoubleValue("MyDanglingLine", "p_pu"), 1.04999999999969850784);
```
Comment (gautierbureau) [inline dynawo/sources/Modeler/DataInterface/test/TestIIDMModels.cpp]: 1.05 not working ?

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelDanglingLine.cpp [hunk 9]
```diff
@@ -712,7 +712,7 @@ ModelDanglingLine::evalJCalculatedVarI(int numCalculatedVar, double* y, double*
   switch (numCalculatedVar) {
     case iNum_: {
       double I = sqrt(Ii1 * Ii1 + Ir1 * Ir1);
-      if (getConnectionState() == CLOSED && I != 0) {
+      if (getConnectionState() == CLOSED && doubleNotEquals(I, 0)) {
```
Comment (gautierbureau) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelDanglingLine.cpp]: 0.

File: dynawo/sources/API/CRV/CRVCurve.cpp (hunk 1 above)
Comment (rosiereflo) [reply dynawo/sources/API/CRV/CRVCurve.cpp]: Not needed there is an embedded test in delete

**Pull request merged**